src/storage/mod.rs

```diff
@@ -0,0 +1,156 @@
+use thiserror::Error;
+
+mod nats_kv;
+mod state;
+
+use semver::Version;
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use state::{
+    LatticeState,
+};
+
+pub use nats_kv::{
+    NatsKvStorageEngine,
+    NatsKvStorageConfig,
+};
+pub use state::{
+    LatticeStateBuilder,
+    LatticeParameters,
+    Host,
+    HostBuilder,
+    Actor,
+    ActorBuilder,
+    Provider,
+    ProviderBuilder,
+    Claim,
+    ClaimBuilder,
+};
+
+/////////////
+// Engines //
+/////////////
+
+/// Pluggable engines can be asked for their metadata
+/// to enable generic (possibly dyn) usage
+#[async_trait]
+pub trait EngineMetadata {
+    /// Get the name of the lattice storage
+    async fn name() -> String;
+
+    /// Get the version of the lattice storage
+    async fn version() -> Version;
+}
+
+/////////////////////
+// Storage Engines //
+/////////////////////
+
+/// Errors that are related to storage engines
+#[derive(Debug, Error)]
+pub enum StorageEngineError {
+    /// When the underlying engine throws an error
+    #[error("Storage engine encountered an error: {0}")]
+    Engine(String),
+
+    #[error("Storage engine [{0}] not connected")]
+    NotConnected(String),
+
+    /// When an operation is not implemented
+    #[error("Operation not implemented: {0}")]
+    NotImplemented(String),
+
+    /// An error when the cause is unknown. This happens very rarely and should be considered fatal
+    #[error("Unknown error has occured")]
+    Unknown,
+}
+
+/////////////
+// Storage //
+/////////////
+
+/// Errors that are related to storage
+#[derive(Debug, Error)]
+pub enum StorageError {
+    /// Errors that arise at the engine layer
+    #[error("Storage error: {0}")]
+    Engine(String),
+
+    /// Errors that arise at the network layer
+    #[error("Network error: {0}")]
+    NetworkError(String),
+
+    /// An error when the cause is unknown. This happens very rarely and should be considered fatal
+    #[error("Unknown error has occured: {0}")]
+    Unknown(String),
+}
+
+/// Options and metadata that can be used to improve store operations
+#[derive(Debug, Serialize, Deserialize)]
+#[derive(Default)]
+pub struct StoreOptions {
+    /// Whether to force a store with compare-and-set semantics
+    pub require_cas: bool,
+
+    /// version identifier that enables stores to perform compare-and-set semantics
+    pub revision: Option<u64>,
+}
+
+
+
+/// A trait that represents whether some struct can be read from a given store
```

Comment:
This isn't something that needs to be changed, but is more of an architectural design preference for evolutionary (rather than predictive) architectures. Here it feels like we've made this code a bit more difficult to read in order to support multiple storage engines in the future. In the interest of the  "keep it simple" principle the code would probably be simpler if we just had a single storage engine for now and extrapolated to generic types later as needed.

Again, not something that I think prevents the PR from being accepted, just want to be explicit that we're incurring some readability debt here.

Reply:
I think this *can* be somewhat mitigated with a file like `docs/storage/writing-a-storage-engine.md` or something of the sort (which has a minimal example), but yeah, I definitely agree with the loss of readability here...

Reply:
After looking at the full NATS implementation, I am inclined to agree with @autodidaddict here that it makes things harder to read without adding much value for us at this point in time. I feel like it might be simpler to just do what we need. I offer the following compromise: What if we changed this so it was a single `Store` trait with all the needed methods (still allows for additional implementations if someone needs it) and then had the NATS impl for that trait?